doc(backup): multiple backup stores support (#1035)
longhorn/website: content/docs/1.8.0/snapshots-and-backups/backup-and-restore/set-backup-target.md

weight: 1
---

> A backup target is an endpoint used to access a backupstore in Longhorn. A backupstore is an NFS server, SMB/CIFS server, Azure Blob Storage server, or S3 compatible server that stores the backups of Longhorn volumes. The backup target can be set at `Settings/Backup Target`.

Author: Modified. Thanks

{{< figure alt="the backup target UI page" src="/img/screenshots/backup-target/page.png" >}}

> > **Note:** Staring with v1.8.0, Longhorn supports the multiple backup stores. We recommend setting the default backup target first, before creating a new one.

Author: Modified. Thanks

- [Set up SMB/CIFS Backupstore](#set-up-smbcifs-backupstore)
- [Set up Azure Blob Storage Backupstore](#set-up-azure-blob-storage-backupstore)

> ### Set up The Default Backup Target
> 
> The default backup target, `default`, will automatically create during a fresh installation. You can manage the default backup target during or after the installation with Helm or the deploying manifest YAML file [longhorn.yaml](https://raw.githubusercontent.com/longhorn/longhorn/v1.8.0/deploy/longhorn.yaml).

Reviewer: ```suggestion
### Default Backup Target

The default backup target (`default`) is automatically created during a fresh installation. You can set the default backup target during or after the installation using either Helm or a [manifest YAML file](https://raw.githubusercontent.com/longhorn/longhorn/v1.8.0/deploy/longhorn.yaml)(`longhorn.yaml`).
```

Author: Modified. Thanks

> #### Set up With Helm
> 
> In the values.yaml, you can set three parameters to manage the default backup target:
> 
> - `defaultBackupStore.backupTarget`: It's the URL used to access the default backupstore.
> - `defaultBackupStore.backupTargetCredentialSecret`: The Kubernetes secret associated with the default backup target if necessary.
> - `defaultBackupStore.pollInterval`: Polling interval to get the backup information from the default backupstore.

Reviewer: ```suggestion
#### Set the Default Backup Target Using Helm

In the `values.yaml` file, you can set three parameters to manage the default backup target.

- `defaultBackupStore.backupTarget`: Endpoint used to access the default backupstore.
- `defaultBackupStore.backupTargetCredentialSecret`: Name of the Kubernetes secret associated with the default backup target.
- `defaultBackupStore.pollInterval`: Number of seconds that Longhorn waits before checking the default backupstore for new backups.
```

Author: Modified. Thanks

```yaml
# -- Setting that allows you to update the default backupstore.
defaultBackupStore:
  # -- Endpoint used to access the default backupstore.
  backupTarget: ~
  # -- Name of the Kubernetes secret associated with the default backup target.
  backupTargetCredentialSecret: ~
  # -- Number of seconds that Longhorn waits before checking the default backupstore for new backups.
  pollInterval: ~
```

> #### Set up With Deploying Manifest YAML File
> 
> Starting with v1.8.0, a new `ConfigMap` named `longhorn-default-resource` is introduced to manage the resources' settings including the default backup target resource:
> 
> - `backup-target`: It's the URL used to access the default backupstore.
> - `backup-target-credential-secret`: The Kubernetes secret associated with the default backup target if necessary.
> - `backupstore-poll-interval`: Polling interval to get the backup information from the default backupstore.

Reviewer: ```suggestion
#### Set the Default Backup Target Using a Manifest YAML File

Starting with v1.8.0, you can use a new `ConfigMap` resource named `longhorn-default-resource` to manage settings of resources, including the default backup target resource.

- `backup-target`: Endpoint used to access the default backupstore.
- `backup-target-credential-secret`: Name of the Kubernetes secret associated with the default backup target.
- `backupstore-poll-interval`: Number of seconds that Longhorn waits before checking the default backupstore for new backups.
```

Author: Modified. Thanks

           -n longhorn-system
       ```

> 3. Go to the Longhorn UI. In the top navigation bar, click **Setting/Backup Target**, and create or edit a backup target.

Author: Modified. Thanks

- The Secret will be mapped to the `AWS_SECRET_ACCESS_KEY` field in the Kubernetes secret we create later.
- The Storage URI will be mapped to the `AWS_ENDPOINTS` field in the Kubernetes secret we create later.

> 9. Go to the Longhorn UI. In the top navigation bar, click **Setting/Backup Target**, and create or edit a backup target as [setting up the AWS S3 backupstore](#set-up-aws-s3-backupstore).

Author: Removed `as [setting up the AWS S3 backupstore](#set-up-aws-s3-backupstore)` first.
It describes that the action is the same as the [item/action 3 in #set-up-aws-s3-backupstore](https://github.com/longhorn/website/pull/1035/files#diff-5df95bd80a43dc3880ea479a2fc2d9aeebc2cf0504101f4f7811937ecd9713f7R142).
Any suggestion to refine it or remove it?

Set **URL** to:

```
s3://${BUCKET_NAME}@us/
```

> And set **Credential Secret** to:

Author: Modified. Thanks

    kubectl create -f https://raw.githubusercontent.com/longhorn/longhorn/v{{< current-version >}}/deploy/backupstores/minio-backupstore.yaml
    ```

> 2. Go to the Longhorn UI. click **Setting/Backup Target**, and create or edit a backup target as [setting up the AWS S3 backupstore](#set-up-aws-s3-backupstore).

Author: Removed `as [setting up the AWS S3 backupstore](#set-up-aws-s3-backupstore)` first. Thanks.

  Set **URL** to:

    ```
    s3://backupbucket@us-east-1/
    ```

>    And set **Credential Secret** to:

Author: Modified. Thanks

  kubectl apply -f cifs_secret.yml
  ```

> Then, navigate to Longhorn UI > Setting > Backup Target

Author: Modified. Thanks

> 1. Create or edit a backup target as [setting up the AWS S3 backupstore](#set-up-aws-s3-backupstore).

Author: Removed `as [setting up the AWS S3 backupstore](#set-up-aws-s3-backupstore)` first. Thanks.

	Any mount options that you specify will replace, not add to, the default options.

>   Set **Credential Secret**

Author: Modified. Thanks

### Set up Azure Blob Storage Backupstore

1. Ensure there is a container used for the backupstore in [Azure Blob Storage Service](https://portal.azure.com/).
2. Give the Azure serviceaccount permissions to read, write, and delete objects in the container.  
  Here is a reference to create a container and grant access to a serviceaccount  

>   https://learn.microsoft.com/en-us/azure/storage/blobs/blob-containers-portal.

Author: Modified. Thanks

3. Navigate to Home > `serviceaccount` > Security + networking > Access keys

> 4. Save the `Storage account name` and `Key`.

Author: Modified. Thanks

- The `Storage account name` will be mapped to the `AZBLOB_ACCOUNT_NAME` field in the Kubernetes secret we create later.

> - The `Key` will be mapped to the `AZBLOB_ACCOUNT_KEY` field in the Kubernetes secret we create later.

Author: Modified. Thanks

> 5. Go to the Longhorn UI. In the top navigation bar, click **Setting/Backup Target**, and create or edit a backup target as [setting up the AWS S3 backupstore](#set-up-aws-s3-backupstore).

Author: Removed `as [setting up the AWS S3 backupstore](#set-up-aws-s3-backupstore)` first. Thanks.

  Set **URL**. The target URL should look like this:

    ```txt
    azblob://[your-container-name]@core.windows.net/
    ```

   Make sure that you have `/` at the end, otherwise you will get an error. A subdirectory (prefix) may be used:

    ```text
    azblob://[your-container-name]@core.windows.net/my-path/
    ```

  Set **Credential Secret**.

    ```txt
    longhorn-azblob-secret
    ```

> 6. Create a Kubernetes secret named `longhorn-azblob-secret` used to access the backupstore in the Longhorn namespace (default is `longhorn-system`) with the following content:

Reviewer: ```suggestion
6. Create a Kubernetes secret named `longhorn-azblob-secret`.

  This secret is used to access the backupstore in the Longhorn namespace (default: `longhorn-system`) with the following content:
```

Author: Modified. Thanks

4. Save the following information:

  - `Storage account name`: Maps to the `AZBLOB_ACCOUNT_NAME` field in the Kubernetes secret that you will create.
  - `Key`: Maps to the `AZBLOB_ACCOUNT_KEY` field in the Kubernetes secret that you will create.

5. Go to the Longhorn UI. In the top navigation bar, click **Setting/Backup Target**, and create or edit a backup target.

  Set **URL**. The target URL should look like this:

    ```txt
    azblob://[your-container-name]@core.windows.net/
    ```

   Make sure that you have `/` at the end, otherwise you will get an error. A subdirectory (prefix) may be used:

    ```txt
    azblob://[your-container-name]@core.windows.net/my-path/
    ```

  Set **Credential Secret**.

    ```txt
    longhorn-azblob-secret
    ```

6. 6. Create a Kubernetes secret named `longhorn-azblob-secret`.
  This secret is used to access the backupstore in the Longhorn namespace (default: `longhorn-system`) with the following content:

  ```shell
  #!/bin/bash
  cat <<EOF >>longhorn-azblob-secret.yml
  apiVersion: v1
  kind: Secret
  metadata:
    name: longhorn-azblob-secret
    namespace: longhorn-system
  type: Opaque
  stringData:

>     AZBLOB_ACCOUNT_NAME: "<Storage account name>"

Reviewer 2: Don't you need to mention the following configurations anymore?

```yaml
      #AZBLOB_ENDPOINT: ${AZBLOB_ENDPOINT}
      #AZBLOB_CERT: ${AZBLOB_CERT}
      #HTTP_PROXY: aHR0cDovLzEwLjIxLjkxLjUxOjMxMjg=
      #HTTPS_PROXY: aHR0cDovLzEwLjIxLjkxLjUxOjMxMjg=
```

Author: Listing these parameters is a bit confusing to users
Using `AZBLOB_ACCOUNT_NAME` and `AZBLOB_ACCOUNT_KEY` parameters is enough to access the Azure blob storage server.
Added the parameters back and added the description to explain how and when to use the parameters.
Thanks
